src/api.ts

```diff
@@ -474,10 +488,159 @@ export async function proxyRequestToHash( req: any, res: any ) {
 		return;
 	}
 
+	touchCommit( commitHash, runEnv );
 	proxy.web( req, res, { target: `http://localhost:${ port }` }, err => {
 		if ( err && ( err as any ).code === 'ECONNRESET' ) {
 			return;
 		}
 		l.log( { err, req, res, commitHash }, 'unexpected error occured while proxying' );
 	} );
 }
+
+export function getContainerName( container: ContainerInfo ) {
+	// The first character is a `/`, skip it
+	return container.Names[ 0 ].substring( 1 );
+}
+
+export function findContainer( { id, image, env, status, name }: ContainerSearchOptions ) {
+	return Array.from( state.containers.values() ).find( container => {
+		if ( image && ( container.Image !== image && container.ImageID !== image ) ) return false;
+		if ( env && container.Labels[ 'calypsoEnvironment' ] !== env ) return false;
+		if ( status && container.Status !== status ) return false;
+		if ( id && container.Id !== id ) return false;
+		// In the Docker internal list, names start with `/`
+		if ( name && ! container.Names.includes( '/' + name ) ) return false;
+		return true;
+	} );
+}
+
+export async function proxyRequestToContainer( req: any, res: any, container: ContainerInfo ) {
+	// In the Docker internal list, names start with `/`
+	const containerName = getContainerName( container );
+
+	if ( ! container.Ports[ 0 ] ) {
+		l.log( { containerName }, `Could not find port for container` );
+		throw new Error( `Could not find port for container ${ containerName }` );
+	}
+	const port = container.Ports[ 0 ].PublicPort;
+
+	let retryCounter = config.proxyRetry;
+	const proxyToContainer = () =>
+		proxy.web( req, res, { target: `http://localhost:${ port }` }, errorHandler );
+	const errorHandler = ( err: any ) => {
+		if ( err && ( err as any ).code === 'ECONNRESET' ) {
```

Comment:
I'm not sure why the request fails with `ECONNRESET` some times, but it is quite frequent. The effect for the user is they go to `http://calypso.live?image=...`, and after downloading the image the browser stays "Waiting for http://container-xxx.calypso.live/". Refreshing the page fixes it.

I _suspect_ is related to my local Docker setup and my laptop's CPU power (or lack thereof), but I'm not sure. Let's see if this also happens in prod.